tests/Highlighter.py

```diff
@@ -0,0 +1,358 @@
+
+    # apply rules on code
+    def highlight(self, code):
+        outputs = []
+        token = "start"
+        for line in code.split("\n"):
+            output, token = self.highlight_rules_line(line, token)
+            outputs.append(output)
+        return "\n".join(outputs)
+
+    # apply rule in one line of code
+    def highlight_rules_line(self, code, start_token="start"):
+        # Initialisation output
+        output = []
+
+        # Initialisation variables
+        current_state = start_token
+        current_position = 0
+
+        default_token = "text"
+
+        end_of_search = False
+        while not end_of_search:
+
+            find_transition, next_transition = self.find_match(code, current_state, current_position)
+
+            if find_transition :
+                # Application of coloring on the code
+
+                # get match
+                current_rule, current_match = next_transition["rule"], next_transition["match"]
+
+                # we color the characters since the last match with the default coloring
+                for c in range(current_position, current_match.start()):
+                    output.append(TOKEN_CODE[default_token])
+
+                # we recover the next default coloring
+                if "defaultToken" in current_rule:
+                    default_token = current_rule["defaultToken"]
+                else:
+                    default_token = "text"
+
+        
+                # if rule has only one groups
+                if current_rule["nb_groups"] == 0:
+                    tok    = current_rule['token']
+                    start  = current_match.start()
+                    length = current_match.end() - current_match.start()
+                    output.append(TOKEN_CODE[tok] * length)
+
+                # if rule has only multiples groups
+                else:
+                    pos = current_match.start()
+                    for i, submatch in enumerate(current_match.groups()):    
+                        tok = current_rule['token'][i%len(current_rule['token'])]        
+                        output.append(TOKEN_CODE[tok] * len(submatch))
+                        pos += len(submatch)
+
+                current_position = current_match.end()
+
+                if 'next' in current_rule:
+                    current_state = current_rule['next']
+
+
+                if current_position == len(code):
+                    end_of_search = True
+
+            else:
+                end_of_search = True
+
+        # we color the last characters since the last match with the default coloring
+        for c in range(current_position, len(code)):
+            output.append(TOKEN_CODE[default_token])
+
+        return "".join(output), current_state
+
+    # find next match
+    def find_match(self, code, current_state, current_position):
```

Comment:
Please document your functions a little in the style as described [here](https://peps.python.org/pep-0008/#documentation-strings). This will make it a lot easier to review and also reason about it:

For example:

```py
def find_match(self, code, current_state, current_position):
  """Find the next matching rule in the given code.

   If there are multiple rules that match the code in the given state, returns the
   one that matches earliest in the source string.

   Arguments:
     - code (string): the code we're searching
     - current_state (string): name of the highlight state we are in
     - current_position (int): index to start searching

   Returns:
     (did_match, transition): a tuple of a boolean and a transition dictionary. The 
        transition dictionary contains  {rule, match} fields and should only be inspected
        if did_match is True.
  """
```

If you wrote this, we can reflect and say: why does this even need to return a tuple at all?
Why doesn't it return an *optional* Transition? That's even simpler to work with for callers.

```py
  """
   Returns:
     {rule, match} | None: a transition dictionary if a match was found, None otherwise.
  """
```

Reply:
Also I think the argument order makes most sense to put `code` and `current_position` next to each other, and `current_state` later. Conceptually those 2 belong together.

Reply:
Not trying to force you to change anything, just giving you something to think about.